### PR TITLE
RSS Виртуальная клавиатура

### PR DESCRIPTION
**Pull request by Asindeton**
1. Задание - https://github.com/rolling-scopes-school/tasks/blob/master/tasks/codejam-virtual-keyboard.md
2. Скриншот - https://imgur.com/tUoP8Ge
3. 06.11.2019 / 06.11.2019
4. Total
 минимальный набор:

    

- [x] реализована генерация DOM-элементов и index.html пустой изначально: +20
- [x]  нажатие на кнопку на физической клавиатуре подсвечивает кнопку на виртуальной: +10

стандартный набор:

  - [x]  есть переключение между русским и английским языком, а так же <s>сохранение выбранного языка</s>: +15
   - [x] клики по кнопкам на виртуальной клавиатуре выводят символы в инпут: +15

повышенный набор:

 - [x] реализована анимация нажатия на кнопку: +15

технические требования:

  - [x]  использование в коде фишек стандарта ES6 и выше (classes, деструктуризацию и тд): +15
  - [x]  использование eslint: +10
  - [x] требования к репозиторию, коммитам и PR выполены: +10

mark calculation:
20+10+10+15+15+10+10 = 90

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  "devDependencies": {
+    "@typescript-eslint/eslint-plugin": "^2.6.1",
+    "eslint": "^6.1.0",
+    "eslint-config-airbnb": "^18.0.1",
+    "eslint-config-airbnb-base": "^14.0.0",
+    "eslint-plugin-import": "^2.18.2",
+    "eslint-plugin-jsx-a11y": "^6.2.3",
+    "eslint-plugin-react": "^7.14.3",
+    "eslint-plugin-react-hooks": "^1.7.0",
+    "eslint-plugin-vue": "^5.2.3"
```
Comment (belashjack) [inline package.json:16]: You tried really hard to install `eslint`. Tried all the options) But you needed only 2 of these.

File: package.json [hunk 2]
```diff
@@ -0,0 +1,31 @@
+{
+  "name": "virtual_keyboard",
+  "version": "1.0.0",
+  "description": "$ npm install -g sass",
```
Comment (belashjack) [inline package.json:4]: Strange description

File: .eslintrc.js [hunk 3]
```diff
@@ -0,0 +1,267 @@
+module.exports = {
```
Comment (belashjack) [inline .eslintrc.js:1]: Weird config for eslint, if you downloaded airbnb config you are expected to use it, but I hope you got the idea how it works after our call

File: style.scss [hunk 4]
```diff
@@ -0,0 +1,120 @@
+$bg_color: #493E3F;
+$bg_color_keyboard: #736962;
+$bg_button: #4C4B40	;
+$bg_button_hover: #2F1417;
+
+
+@font-face {
```
Comment (belashjack) [inline style.scss:7]: Move to separate file

File: style.scss [hunk 5]
```diff
@@ -0,0 +1,120 @@
+$bg_color: #493E3F;
+$bg_color_keyboard: #736962;
+$bg_button: #4C4B40	;
+$bg_button_hover: #2F1417;
```
Comment (belashjack) [inline style.scss:4]: To separate file

File: script.js [hunk 6]
```diff
@@ -0,0 +1,689 @@
+
+let test = '`';
+let test2 = '/';
+
+
+document.body.insertAdjacentHTML("afterbegin", `<div class="wrapper">
+<div class="textArea">
```
Comment (belashjack) [inline script.js:7]: Use `text_area` or `text_area` notation here and everywhere for classes

File: script.js [hunk 7]
```diff
@@ -0,0 +1,689 @@
+
+let test = '`';
+let test2 = '/';
+
+
+document.body.insertAdjacentHTML("afterbegin", `<div class="wrapper">
```
Comment (belashjack) [inline script.js:6]: Well, this looks like a hack. I assume the idea was to really generate html from some data. Generate means "bring to existence from something". You generate nothing, you've just written the same html in another place. You could store the data for your buttons something like:

`const buttons = [
{ value: { en: 'q', ru: 'й' }, code: 81 },
{ value: { en: 'w', ru: 'ц' }, code: 87 },
...
]`

And then generate html from this data:

`buttons.forEach((buttonData) => {
const button = document.createElement('div');
button.textContent = buttonData.value[currentLanguage]; // store currentLanguage somewhere
...
document.body.appendChild(button);
})`

File: script.js [hunk 8]
```diff
@@ -0,0 +1,689 @@
...
+<div class="textArea">
+  <textarea name="textarea" id="name" class="textarea" cols="30" rows="10" autofocus  placeholder="Write something here"></textarea>
+</div>
+<div class="virtualKeyboard">
+  <div class="keyboard_row">
+    <div class="key" id="Backquote">
+      <span class="Back_quote on">
+        <span class="case off">Ё</span>
+        <span class="case on">ё</span>
```
Comment (belashjack) [inline script.js:15]: 1) I can only guess that `on` and `off` relates to `Shift` button. But I MUST understand it from your code. Rename this classes.
2) Why do you need to have `on` in two places? The second is for the language? Then why they are named the same?

File: script.js [hunk 9]
```diff
@@ -0,0 +1,689 @@
...
+      <span class="case arrow_right">►</span>
+    </div>
+    <div class="key" id="ControlRight">
+      <span class="case ctrl">Ctrl</span>
+    </div>
+  </div>
+    </div>       
+</div>`);
+
+let textarea = document.getElementById('name');
```
Comment (belashjack) [inline script.js:596]: `const` here and everywhere it is possible to use instead of `let`

File: .eslintrc.js (hunk 3 above)
Comment (Asindeton) [reply .eslintrc.js:1]: Yes)

File: script.js [hunk 10]
```diff
@@ -0,0 +1,689 @@
...
+let changeLang = document.querySelectorAll('.key>.on');
+let changeLang2 = document.querySelectorAll('.key>.off');
+
+document.onkeydown = function(e){
```
Comment (belashjack) [inline script.js:600]: `addEventListener` !!!!! Here and everywhere

File: script.js [hunk 11]
```diff
@@ -0,0 +1,689 @@
...
+  let capsButton = document.querySelectorAll(".on>.case");
+  let activeKey = document.getElementsByClassName("key");
```
Comment (belashjack) [inline script.js:602]: `querySelector`!!! Here and everywhere

Comment (belashjack) [inline script.js:602]: Use plural form for arrays (`capsButtons`, `activeKeys`), to say that your names are confusing is to say nothing

File: script.js [hunk 12]
```diff
@@ -0,0 +1,689 @@
...
+  textarea.focus();
+    for(let i = 0; i < activeKey.length;i++){
+      if (activeKey[i].id == e.code){
```
Comment (belashjack) [inline script.js:605]: Use `===` always

File: script.js [hunk 13]
```diff
@@ -0,0 +1,689 @@
...
+
+let textarea = document.getElementById('name');
+let changeLang = document.querySelectorAll('.key>.on');
+let changeLang2 = document.querySelectorAll('.key>.off');
+
+document.onkeydown = function(e){
+  let capsButton = document.querySelectorAll(".on>.case");
+  let activeKey = document.getElementsByClassName("key");
+  textarea.focus();
+    for(let i = 0; i < activeKey.length;i++){
```
Comment (belashjack) [inline script.js:604]: You could use `Array.find()`

File: script.js [hunk 14]
```diff
@@ -0,0 +1,689 @@
...
+      if (activeKey[i].id == e.code){
+        if (e.code == 'ShiftLeft'){
+          for (let i = 0; i < capsButton.length; i++){
```
Comment (belashjack) [inline script.js:607]: Hell no. For loops inside loops don't reuse `i`, there are 25 letters left, usually people use `j` and `k`

File: script.js [hunk 15]
```diff
@@ -0,0 +1,689 @@
...
+    <div class="key" id="ControlRight">
+      <span class="case ctrl">Ctrl</span>
+    </div>
+  </div>
+    </div>       
+</div>`);
+
+let textarea = document.getElementById('name');
+let changeLang = document.querySelectorAll('.key>.on');
+let changeLang2 = document.querySelectorAll('.key>.off');
```
Comment (belashjack) [inline script.js:598]: What is `changeLang`? I from the name I expect it is a function, but it is not. What is changeLang2? What is 2???? It is just unreadable and usually it means that code is designed poorly

File: script.js [hunk 16]
```diff
@@ -0,0 +1,689 @@
...
+        }
+     }
+ }
+
+
+ document.onmousedown = function(){
+   let presButton = document.querySelectorAll(":hover>.on>.case.on");
+   let capsButton = document.querySelectorAll(".on>.case");
+  for(let i=0; i < presButton.length; i++){
+     if(presButton[i].innerText == ""){
```
Comment (belashjack) [inline script.js:658]: Prefer `switch` to such long `if-else` statements`

**Pull request closed without merging**